drivers/opencl/src/main/java/uk/ac/manchester/tornado/drivers/opencl/graal/nodes/vector/VectorValueNode.java

```diff
@@ -160,30 +159,53 @@ private void generateVectorAssign(NodeLIRBuilderTool gen, LIRGeneratorTool tool,
 
         Value s0,s1,s2,s3,s4,s5,s6,s7;
         switch (kind.getVectorLength()) {
-            case 2:
+            case 2: {
                 final OCLOp2 op2 = VectorUtil.resolveAssignOp2(getOCLKind());
                 s0 = getParam(gen, tool, 0);
+                OCLKind kindOp = (OCLKind) s0.getPlatformKind();
+                if (kindOp.isVector()) {
+                    gen.setResult(this, s0);
+                    return;
+                }
                 s1 = getParam(gen, tool, 1);
                 assignExpr = new OCLVectorAssign.Assign2Expr(op2, getOCLKind(), s0, s1);
                 break;
-            case 3:
+            }
+            case 3: {
                 final OCLOp3 op3 = VectorUtil.resolveAssignOp3(getOCLKind());
                 s0 = getParam(gen, tool, 0);
+                OCLKind kindOp = (OCLKind) s0.getPlatformKind();
+                if (kindOp.isVector()) {
+                    gen.setResult(this, s0);
+                    return;
+                }
                 s1 = getParam(gen, tool, 1);
                 s2 = getParam(gen, tool, 2);
                 assignExpr = new OCLVectorAssign.Assign3Expr(op3, getOCLKind(), s0, s1, s2);
                 break;
-            case 4:
+            }
+            case 4: {
                 final OCLOp4 op4 = VectorUtil.resolveAssignOp4(getOCLKind());
                 s0 = getParam(gen, tool, 0);
+                OCLKind kindOp = (OCLKind) s0.getPlatformKind();
+                if (kindOp.isVector()) {
+                    gen.setResult(this, s0);
+                    return;
+                }
                 s1 = getParam(gen, tool, 1);
                 s2 = getParam(gen, tool, 2);
                 s3 = getParam(gen, tool, 3);
                 assignExpr = new OCLVectorAssign.Assign4Expr(op4, getOCLKind(), s0, s1, s2, s3);
                 break;
-            case 8:
+            }
+            case 8: {
                 final OCLOp8 op8 = VectorUtil.resolveAssignOp8(getOCLKind());
                 s0 = getParam(gen, tool, 0);
+                OCLKind kindOp = (OCLKind) s0.getPlatformKind();
+                if (kindOp.isVector()) {
+                    gen.setResult(this, s0);
+                    return;
+                }
```

Comment:
Can you provide some explanation here? 

Why is only `s0` used to set the result in the `NodeLIRBuilderTool`?

Reply:
Yes. So what happens is that the JIT compiler tried to create a new vector type from scalar values. However, we need to cover the case in which we just make a copy of a vector type (without invoking the subcomponents)

```
vector4 a = b;
```

Otherwise, the compiler will generate the following sequence (which is incorrect)

```
vector4 a = (vector4) (b, 0, 0, 0);  // b is vector4
```

For those cases, we need to control de `PlatformKind` of the first operand. In the case is a vector, we just pass the result to the next `LIROp` operation.

Reply:
yes, it makes sense. Thanks for the info.